app/main.py

```diff
@@ -1,2 +1,70 @@
+from typing import Any, Hashable
+
+
 class Dictionary:
-    pass
+    class Node:
+        def __init__(self, key: Hashable, value: Any) -> None:
+            self.key = key
+            self.value = value
+
+    def __init__(self) -> None:
+        self.capacity = 8
+        self.size = 0
+        self.load_factor_threshold = 2 / 3
+        self.hash_table: list[tuple | None] = [None] * self.capacity
+
+    def calculate_hash(self, key: Hashable) -> int:
+        index = hash(key) % self.capacity
+
+        while (
+            self.hash_table[index] is not None
+            and self.hash_table[index].key != key
+        ):
+            index = (index + 1) % self.capacity
+        return index
+
+    @property
+    def current_max_size(self) -> int:
+        return int(self.capacity * self.load_factor_threshold)
+
+    def resize(self) -> None:
+        old_hash_table = self.hash_table
+        self.capacity *= 2
+        self.hash_table = [None] * self.capacity
+        self.size = 0
+
+        for node in old_hash_table:
+            if node is not None:
+                self.__setitem__(node.key, node.value)
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
+        index = self.calculate_hash(key)
+
+        if self.hash_table[index] is None:
+            if self.size + 1 >= self.current_max_size:
+                self.resize()
+                return self.__setitem__(key, value)
+            self.size += 1
+
+        self.hash_table[index] = Dictionary.Node(key, value)
+
+    def __getitem__(self, key: Hashable) -> Any:
+        index = self.calculate_hash(key)
+        if self.hash_table[index] is None:
+            raise KeyError(f"No value for key: {key}")
+        return self.hash_table[index].value
+
+    def __len__(self) -> int:
+        return self.size
+
+    def clear(self) -> None:
+        self.hash_table = [None] * self.capacity
+        self.size = 0
+
+    def __delitem__(self, key: Hashable) -> None:
+        index = self.calculate_hash(key)
+        if self.hash_table[index] is None:
+            raise KeyError(f"No value for key: {key}")
+
+        self.hash_table[index] = None
+        self.size -= 1
```

Comment:
you need to recalculate remaining positions after deleting an element because indexes may be wrong for elements that had collisions